ir/src/symbol_table.rs

```diff
@@ -0,0 +1,178 @@
+            Self::AuxTraceColumn(_) => "AuxTraceColumn",
+        };
+
+        write!(f, "{}", display)
+    }
+}
+
+/// SymbolTable for identifiers to track their types and information and enforce uniqueness of
+/// identifiers.
+#[derive(Default, Debug)]
+pub(super) struct SymbolTable {
+    /// A map of all declared identifiers from their name (the key) to their type.
+    identifiers: BTreeMap<String, IdentifierType>,
+
+    /// A map of the Air's periodic columns using the index of the column within the declared
+    /// periodic columns as the key and the vector of periodic values as the value
+    periodic_columns: Vec<Vec<u64>>,
+}
+
+impl SymbolTable {
+    /// TODO
+    fn insert_symbol(
+        &mut self,
+        ident_name: &str,
+        ident_type: IdentifierType,
+    ) -> Result<(), SemanticError> {
+        let result = self.identifiers.insert(ident_name.to_owned(), ident_type);
+        match result {
+            Some(prev_type) => Err(SemanticError::DuplicateIdentifier(format!(
+                "Cannot declare {} as a {}, since it was already defined as a {}",
+                ident_name, ident_type, prev_type
+            ))),
+            None => Ok(()),
+        }
+    }
+
+    // --- MUTATORS -------------------------------------------------------------------------------
+
+    /// TODO Add a new main trace column by its name and index in the main execution trace.
+    pub(super) fn insert_main_trace_columns(
+        &mut self,
+        columns: &[Identifier],
+    ) -> Result<(), SemanticError> {
+        for (idx, Identifier(name)) in columns.iter().enumerate() {
+            self.insert_symbol(name, IdentifierType::MainTraceColumn(idx))?;
+        }
+
+        Ok(())
+    }
+
+    /// TODO Add a new auxiliary trace column by its name and index in the auxiliary execution trace.
+    pub(super) fn insert_aux_trace_columns(
+        &mut self,
+        columns: &[Identifier],
+    ) -> Result<(), SemanticError> {
+        for (idx, Identifier(name)) in columns.iter().enumerate() {
+            self.insert_symbol(name, IdentifierType::AuxTraceColumn(idx))?;
+        }
+
+        Ok(())
+    }
+
+    /// Adds a new public input by its name and size.
+    pub(super) fn insert_public_inputs(
+        &mut self,
+        public_inputs: &[PublicInput],
+    ) -> Result<(), SemanticError> {
+        for input in public_inputs.iter() {
+            self.insert_symbol(input.name(), IdentifierType::PublicInput(input.size()))?;
+        }
+
+        Ok(())
+    }
+
+    /// TODO
+    pub(super) fn insert_periodic_columns(
+        &mut self,
+        columns: &[PeriodicColumn],
+    ) -> Result<(), SemanticError> {
+        for (index, column) in columns.iter().enumerate() {
+            validate_cycles(column)?;
+
+            self.insert_symbol(column.name(), IdentifierType::PeriodicColumn(index))?;
+            self.periodic_columns.push(column.values().to_vec());
+        }
+
+        Ok(())
```

Comment:
Comment for the future: what should we do if two periodic columns are identical (same cycle length and values)? Should it be an error? A warning? Or something else?

Reply:
Good point. Probably a warning? But we could ask also fix it internally

Reply:
I've opened #31